text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,648 @@
+  constructor(scope: Construct, id: string, props: OriginAccessControlProps = {}) {
+    super(scope, id);
+
+    const resource = new CfnOriginAccessControl(this, 'Resource', {
+      originAccessControlConfig: {
+        description: props.description,
+        name: props.originAccessControlName ?? this.generateName(),
+        signingBehavior: props.signingBehavior ?? SigningBehavior.ALWAYS,
+        signingProtocol: props.signingProtocol ?? SigningProtocol.SIGV4,
+        originAccessControlOriginType: props.originAccessControlOriginType ?? OriginAccessControlOriginType.S3,
+      },
+    });
+
+    this.originAccessControlId = resource.attrId;
+  }
+
+  private generateName(): string {
+    const name = Stack.of(this).region + Names.uniqueId(this);
+    if (name.length > 64) {
+      return name.substring(0, 32) + name.substring(name.length - 32);
+    }
+    return name;
+  }
+
+}
+```
+
+#### New `S3BucketOacOrigin` class
+
+A new class `S3BucketOacOrigin` will implement the `bind()` method to setup the OAC and update the bucket policy.
+
+In the case where an imported bucket is being used for the S3 origin, calling `bucket.addToResourcePolicy()` will fail to add the policy statement. Existing
+[workarounds](https://github.com/aws/aws-cdk/issues/6548#issuecomment-869091553) require the user to create a new `BucketPolicy` for the bucket and
+add the policy statements using `bucketPolicy.document.addStatements()`.
+However, this overwrites the whole bucket policy instead of appending statements to the
+existing policy which is a subpar user experience. The proposed solution to this issue is
+to use a custom resource to retrieve the existing bucket policy and append the
+OAC policy statement via the `GetBucketPolicy()` and `PutBucketPolicy()` API calls
+after the CloudFront distribution has been created.
+
+In the case where the S3 bucket uses SSE-KMS encryption (customer-managed key),
+a circular dependency error occurs when trying to deploy the template. When granting
+the CloudFront distribution access to use the KMS Key, there is a circular dependency:
+
+- CloudFront distribution references the S3 bucket
+- S3 bucket references the KMS key
+- KMS Key references the CloudFront distribution
+
+The proposed solution to this issue is to use a custom resource
+to retrieve and update the KMS key policy after the CloudFront
+distribution has been created via the `GetKeyPolicy()` and `PutKeyPolicy()` API calls.
+
+```ts
+class S3BucketOacOrigin extends cloudfront.OriginBase {
+  private originAccessControl!: cloudfront.IOriginAccessControl;
+
+  constructor(private readonly bucket: s3.IBucket, { originAccessControl, ...props }: S3OriginProps) {
+    super(bucket.bucketRegionalDomainName, props);
+    if (originAccessControl) {
+      this.originAccessControl = originAccessControl;
+    }
+  }
+
+  public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
```

Comment:
Before diving in more here, there is a lot going on in this function. Let's try to separate some of this out into smaller functions

Reply:
For sure, added some private functions to make this more readable